### PR TITLE
is_union returns ty to avoid computing it twice

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

**Review by oli-obk — APPROVED** [review]

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 1]
```diff
@@ -632,25 +630,29 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                        if let Some(union_ty) = is_union(base) {
+                            if field != target_field && base == target_base {
+                                let desc_base =
+                                    self.describe_place(base).unwrap_or_else(|| "_".to_owned());
+                                let desc_first = self
+                                    .describe_place(first_borrowed_place)
+                                    .unwrap_or_else(|| "_".to_owned());
+                                let desc_second = self
+                                    .describe_place(second_borrowed_place)
+                                    .unwrap_or_else(|| "_".to_owned());
```
Comment (Centril) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: This is pattern repeated 3 times here... use a lambda?

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 2]
```diff
@@ -595,11 +595,9 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     ) -> (String, String, String, String) {
         // Define a small closure that we can use to check if the type of a place
         // is a union.
-        let is_union = |place: &Place<'tcx>| -> bool {
-            place.ty(self.mir, self.infcx.tcx).ty
-                .ty_adt_def()
-                .map(|adt| adt.is_union())
-                .unwrap_or(false)
+        let is_union = |place: &Place<'tcx>| -> Option<&TyS<'tcx>> {
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: This should be `Option<Ty<'tcx>>`

File: src/librustc_mir/borrow_check/conflict_errors.rs (hunk 1 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/conflict_errors.rs]: Fixed the 4 repetitions :)

File: src/librustc_mir/borrow_check/conflict_errors.rs (hunk 2 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/conflict_errors.rs]: Good catch, fixed.

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 3]
```diff
@@ -632,33 +631,37 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
-                        _ => current = base,
+                    if let ProjectionElem::Field(field, _) = elem {
+                        if let Some(union_ty) = is_union(base) {
+                            if field != target_field && base == target_base {
+                                let desc_base =
+                                    self.describe_place(base).unwrap_or_else(underscore_str_fn);
+                                let desc_first = self
+                                    .describe_place(first_borrowed_place)
+                                    .unwrap_or_else(underscore_str_fn);
+                                let desc_second = self
```
Comment (Centril) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: Hmm; you are still repeating `self.describe_place($some_place).unwrap_or_else(underscore_str_fn)` three times; can't you have:
```
let describe_place = |place| self.describe_place(place).unwrap_or_else(|| "_".to_owned());
```
or something like that and then write:
```rust
return Some((
    describe_place(base),
    describe_place(first_borrowed_place),
    describe_place(second_borrowed_place),
    union_ty.to_string(),
));
```

Comment (spastorino) [reply src/librustc_mir/borrow_check/conflict_errors.rs]: Fixed, I wasn't paying a lot of attention to the existing code to be honest :).

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 4]
```diff
@@ -595,12 +595,11 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     ) -> (String, String, String, String) {
         // Define a small closure that we can use to check if the type of a place
         // is a union.
-        let is_union = |place: &Place<'tcx>| -> bool {
-            place.ty(self.mir, self.infcx.tcx).ty
-                .ty_adt_def()
-                .map(|adt| adt.is_union())
-                .unwrap_or(false)
+        let is_union = |place: &Place<'tcx>| -> Option<Ty<'tcx>> {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: The name of this lambda is suboptimal now that it doesn't return `bool`.

**Review by Centril — APPROVED** [review]

File: src/librustc_mir/borrow_check/conflict_errors.rs (hunk 4 above)
Comment (spastorino) [reply src/librustc_mir/borrow_check/conflict_errors.rs]: Done!

**Comment by spastorino** [comment]
Have also squashed the two last commits into one because it did not make sense to split those.

**Pull request merged**